### ISSUE TITLE
FacetDistribution values are incomplete

### ISSUE THREAD

**Issue by bb** [meilisearch/MeiliSearch#1576]
**Describe the bug**

When doing a search including facetsDistribution in 0.21.0rc4, I don't receive all values for a field, even when I'm not filtering. In earlier versions (0.20 and before) the facetDistribution was complete and correct for the same dataset.

I made sure it's not missing data by checking the values in the hits from the very same response.

My dataset contains a field `region_ids` which is an array of strings (which look like ints, but they're strings because of <v0.21 limitations).

Tiny example (most likely not big enough for reproduction, just to see the structure see [this slack message](https://meilicommunity.slack.com/archives/CP9DVS1RQ/p1628699235091200?thread_ts=1628696972.089500&cid=CP9DVS1RQ) for more):

```json
  [
  { "region_ids": ["-2", "-2202162", "-8649", "-71525"]  },
  { "region_ids": ["-2", "-49715", "-224459", "-2431739"] },
  { "region_ids": ["-5", "-60199", "-71248", "-11900985"] }
]
```

**To Reproduce**
1. Do a search including facets: `res=await index.search("", {facetsDistribution: ["*"], limit: 1000 })`
2. Observe facet results: `Object.keys(res.facetsDistribution.region_ids).length` (18 in my case)
3. Compare with the values in the hits: `new Set(res.hits.map (h => h.region_ids).flat()).size` (44 in my case)
4. facet distribution values (18) should be bigger than number of values in hits (44)

**Expected behavior**
Each field in the facetsDistribution should contain *all* values and their complete counts. Thus the number keys of a field in facetsDistribution should always be greater or equal than the actual values in the hits. If the default behavior is to limit that to a reasonable number, there should be a setting to enable returning all. 

Of course, the facetDistribution should adhere to the filters. It did so for facetFilters in <0.21 but according to #929 not for filters.

**MeiliSearch version:** [e.g. v0.21.0rc4]

**Comment by Kerollmops** [comment]
Hey @bb,

What I understand is that you want the engine to return at least the facet values that are present in the returned documents in the `facetDistribution` field returned. And if that really what you want it is absolutely feasible and that will not impact performances.

@ManyTheFish What we can do to enable this behavior is that we continue to return the values from the inverted index (from the `field_id_facet_string_docids` database) but after we created the HashMap of facet values, we extend it by retrieving the values coming from the documents returned (from the `field_id_docid_facet_string` database). It should not impact the engine's performances as there only is a little amount of documents returned.

**Comment by bb** [comment]
I want *all* facet value counts for all documents (...matching the filter criteria, if any). I just took the returned documents example as a way to proof the situation.

**Comment by curquiza** [comment]
Hello @bb!
We found out we put a hard limit for the facet values. We removed it.
I created a draft branch that should fix your issue -> https://github.com/meilisearch/MeiliSearch/pull/1577 Can you confirm if you have time? 😁
It would be helpful!

**Comment by curquiza** [comment]
However, I want to warn you: the search engine will not return the zero facet value anymore. I mean:
- v0.20.0
```json
"facetsDistribution": {
        "genres": {
            "Action": 10,
            "Adventure": 3,
            "Animation": 0
        }
}
```

- v0.21.0
```json
"facetsDistribution": {
        "genres": {
            "Action": 10,
            "Adventure": 3
        }
}
```

This is impossible for the new search engine to display ALL the facets. This is something we might never change.

Here is a workaround to display them in front-end: https://github.com/meilisearch/meilisearch-js/issues/970

FYI, Algolia does the same, here is their explanation: https://discourse.algolia.com/t/keep-all-filters-even-if-no-results/493

**Comment by bb** [comment]
Thanks for clearing that zero-change up. I wasn't even aware 0.20.0 returns facet values with count zero.

It looks better as the result count is higher than 100, but I still don't have the results I expected:

```
res = await (await fetch("http://localhost:7700/indexes/posts/search", {
  "body": `{"q":"","limit":10000, "attributesToHighlight":["*"],"facetsDistribution":["*"]}`,
  "method": "POST",
})).json();
// {hits: Array(5596), nbHits: 5596, exhaustiveNbHits: false, query: "", limit: 10000, …}
new Set(res.hits.map (h => h.region_ids).flat()).size
// 1133
Object.keys(res.facetsDistribution.region_ids).length
// 252
```

**Comment by bb** [comment]
Some more analysis revealed, that the `facet_numbers_distribution_from_facet_levels` algorithm (which is selected when `nbHits > 3000`) doesn't work for me while the `facet_distribution_from_documents` algorithm is fine. I rebuilt with https://github.com/meilisearch/milli/blob/remove-facet-distribution-limit/milli/src/search/facet/facet_distribution.rs#L18 changed to a number bigger than the count of my documents and now everything works as expected. It'd be great if I could change the threshold via config or opt-in to the `facet_distribution_from_documents` algorithm in some other way (or if the algorithm could be modified to return exact numbers).

**Comment by curquiza** [comment]
@bb, Many just finished the fixed, can you check again.

on the `fix-facets-distribution` branch in the MeiliSearch repository
```
cargo update -p milli
cargo run --release
```

**Comment by bb** [comment]
@curquiza, the fix looks good. Thanks a lot to @many and you!

There was only a change for string facets which ist what I reported and tested. I already chatted with Many in Slack about that, he confirmed that only fields with string arrays were affected.

**Comment by curquiza** [comment]
fixed with #1577
